### PR TITLE
Added fog, horizon, colorful sky

### PR DESCRIPTION
**Pull request by Andrewschenko**
Added features for extended tilt

### CODE REVIEW

File: src/Map/OpenGL/AtlasMapRenderer_OpenGL.cpp [hunk 1]
```diff
@@ -361,6 +341,77 @@ bool OsmAnd::AtlasMapRenderer_OpenGL::updateInternalState(
...
+    groundPos.y = static_cast<double>(tileIdN.y) + offsetInTileN.y;
+    const auto metersPerUnit = Utilities::getMetersPerTileUnit(state.zoomLevel, groundPos.y, TileSize3D);
+    internalState->metersPerUnit = metersPerUnit;
+    internalState->distanceFromCameraToGroundInMeters =
+        qMax(0.0, static_cast<double>(internalState->distanceFromCameraToGround) * metersPerUnit);
+    internalState->cameraCoordinates.x = Utilities::getLongitudeFromTile(state.zoomLevel, groundPos.x);
+    internalState->cameraCoordinates.y = Utilities::getLatitudeFromTile(state.zoomLevel, groundPos.y);
+    
+    // Calculate distance to horizon
+    double radius = 6371e3;
```
Comment (alex-osm) [inline src/Map/OpenGL/AtlasMapRenderer_OpenGL.cpp:364]: Introduce constants for hardcoded numbers like 6371e3

File: src/Map/OpenGL/AtlasMapRendererMapLayersStage_OpenGL.cpp [hunk 2]
```diff
@@ -638,6 +651,13 @@ bool OsmAnd::AtlasMapRendererMapLayersStage_OpenGL::initializeRasterLayersProgra
         "{                                                                                                                  ""\n"
         "    lowp vec4 finalColor;                                                                                          ""\n"
         "                                                                                                                   ""\n"
+        //   Get the distance rate for hiding tiles in mist
+        "    vec4 infrontPosition = v2f_position;                                                                           ""\n"
+        "    infrontPosition.xz = v2f_position.xz * param_fs_mistConfiguration.xy;                                          ""\n"
+        "    infrontPosition.xz = v2f_position.xz - (infrontPosition.x + infrontPosition.z) * param_fs_mistConfiguration.xy;""\n"
+        "    float distanceToCamera = param_fs_mistConfiguration.z-distance(infrontPosition, param_fs_worldCameraPosition); ""\n"
+        "    float notFar = clamp(exp(distanceToCamera * 8.0f / param_fs_mistConfiguration.w - 4.0f), 0.0f, 1.0f);          ""\n"
```
Comment (vshcherb) [inline src/Map/OpenGL/AtlasMapRendererMapLayersStage_OpenGL.cpp:659]: > exp(distanceToCamera * 8.0f / param_fs_mistConfiguration.w - 4.0f
Complex Formula

File: src/Map/OpenGL/AtlasMapRendererSkyStage_OpenGL.cpp [hunk 3]
```diff
@@ -61,12 +68,28 @@ bool OsmAnd::AtlasMapRendererSkyStage_OpenGL::initialize()
...
         "void main()                                                                                                        ""\n"
         "{                                                                                                                  ""\n"
-        "    FRAGMENT_COLOR_OUTPUT = param_fs_skyColor;                                                                     ""\n"
+        "    vec4 color;                                                                                                    ""\n"
+        "    float position;                                                                                                ""\n"
+        "    float skyHeight = param_fs_skySize.y / (1.0f - param_fs_skySize.z);                                            ""\n"
+        "    bool sky = v2f_position.y > param_fs_skySize.z;                                                                ""\n"
+        "    position = (v2f_position.y - (sky ? param_fs_skySize.z : 0.0f)) * skyHeight;                                   ""\n"
+        "    float low = 1.0f / pow(1.1f + position, 4.0f);                                                                 ""\n"
```
Comment (vshcherb) [inline src/Map/OpenGL/AtlasMapRendererSkyStage_OpenGL.cpp:86]: > float low = 1.0f / pow(1.1f + position, 4.0f);       
Complex formula

File: src/Map/OpenGL/AtlasMapRendererSkyStage_OpenGL.cpp [hunk 4]
```diff
@@ -61,12 +68,28 @@ bool OsmAnd::AtlasMapRendererSkyStage_OpenGL::initialize()
...
+        "    float high = 1.0f - low;                                                                                       ""\n"
+        "    color.r = sky ? 0.7647059f * low + high * exp(1.0f - position / 3.0f) * 0.3101728f : param_fs_fogColor.r;      ""\n"
```
Comment (vshcherb) [inline src/Map/OpenGL/AtlasMapRendererSkyStage_OpenGL.cpp:88]: > sky ? 0.7647059f * low + high * exp(1.0f - position / 3.0f) * 0.3101728f 

Complext Formual

File: src/Map/OpenGL/AtlasMapRendererSkyStage_OpenGL.cpp [hunk 5]
```diff
@@ -61,12 +68,28 @@ bool OsmAnd::AtlasMapRendererSkyStage_OpenGL::initialize()
...
+        "    color.g = sky ? 0.8039216f * low + high * exp(1.0f - position / 5.0f) * 0.3390262f : param_fs_fogColor.g;      ""\n"
```
Comment (vshcherb) [inline src/Map/OpenGL/AtlasMapRendererSkyStage_OpenGL.cpp:89]: > 0.7647059f , 0.3101728f,  0.8039216f, 0.3390262f

Looks like color numbers but not in constants

File: src/Map/OpenGL/AtlasMapRenderer_OpenGL.cpp [hunk 6]
```diff
@@ -35,13 +35,28 @@
 
 #include "OpenGL/Utilities_OpenGL.h"
 
+// Set camera's near depth limit
 const float OsmAnd::AtlasMapRenderer_OpenGL::_zNear = 0.1f;
 
+// Set average radius of Earth
+const double OsmAnd::AtlasMapRenderer_OpenGL::_radius = 6371e3;
+// Set minimum earth angle for advanced horizon
+const double OsmAnd::AtlasMapRenderer_OpenGL::_minimumAngleForAdvancedHorizon = M_PI / 12.0;
```
Comment (vshcherb) [inline src/Map/OpenGL/AtlasMapRenderer_OpenGL.cpp:44]: Not clear purpose of constant

File: src/Map/OpenGL/AtlasMapRenderer_OpenGL.cpp [hunk 7]
```diff
@@ -35,13 +35,28 @@
...
+// Set distance-per-angle factor for smooth horizon slide beyond minimum earth angle
+const double OsmAnd::AtlasMapRenderer_OpenGL::_distancePerAngleFactor = 1.1648753803647327974577447574825e-6;
```
Comment (vshcherb) [inline src/Map/OpenGL/AtlasMapRenderer_OpenGL.cpp:46]: Not clear purpose of constant

File: src/Map/OpenGL/AtlasMapRenderer_OpenGL.cpp [hunk 8]
```diff
@@ -35,13 +35,28 @@
...
+// Set minimum height of visible sky for colouring
+const double OsmAnd::AtlasMapRenderer_OpenGL::_minimumSkyHeightInKilometers = 8.0;
```
Comment (vshcherb) [inline src/Map/OpenGL/AtlasMapRenderer_OpenGL.cpp:48]: Not clear purpose of constant

File: src/Map/OpenGL/AtlasMapRenderer_OpenGL.cpp [hunk 9]
```diff
@@ -35,13 +35,28 @@
...
+// Set the minimum height of terrain to render
+const double OsmAnd::AtlasMapRenderer_OpenGL::_minimumHeightFromGroundInMeters = -12000.0;
```
Comment (vshcherb) [inline src/Map/OpenGL/AtlasMapRenderer_OpenGL.cpp:50]: Should we switch to 0 now and later we will enable it if we need.

**Pull request closed without merging**